src/LAMMPS.jl

```diff
@@ -181,240 +236,343 @@ function get_natoms(lmp::LMP)
     Int64(API.lammps_get_natoms(lmp))
 end
 
-function dtype2type(dtype::API._LMP_DATATYPE_CONST)
-    if dtype == API.LAMMPS_INT
-        type = Ptr{Int32}
-    elseif dtype == API.LAMMPS_INT_2D
-        type = Ptr{Ptr{Int32}}
-    elseif dtype == API.LAMMPS_INT64
-        type = Ptr{Int64}
-    elseif dtype == API.LAMMPS_INT64_2D
-        type = Ptr{Ptr{Int64}}
-    elseif dtype == API.LAMMPS_DOUBLE
-        type = Ptr{Float64}
-    elseif dtype == API.LAMMPS_DOUBLE_2D
-        type = Ptr{Ptr{Float64}}
-    elseif dtype == API.LAMMPS_STRING
-        type = Ptr{Cchar}
-    else
-        @assert false "Unknown dtype: $dtype"
-    end
-    return type
+function _lammps_string(ptr::Ptr)
+    ptr == C_NULL && error("Wrapping NULL-pointer!")
+    return Base.unsafe_string(ptr)
 end
 
-"""
-    extract_global(lmp, name, dtype=nothing)
-"""
-function extract_global(lmp::LMP, name, dtype=nothing)
-    if dtype === nothing
-        dtype = API.lammps_extract_global_datatype(lmp, name)
-    end
-    dtype = API._LMP_DATATYPE_CONST(dtype)
-    type = dtype2type(dtype)
+function _lammps_wrap(ptr::Ptr{<:Real}, shape::Integer, copy=true)
+    ptr == C_NULL && error("Wrapping NULL-pointer!")
 
-    ptr = API.lammps_extract_global(lmp, name)
-    ptr = reinterpret(type, ptr)
+    result = Base.unsafe_wrap(Array, ptr, shape, own=false)
 
-    if ptr !== C_NULL
-        if dtype == API.LAMMPS_STRING
-            return Base.unsafe_string(ptr)
-        end
-        # TODO: deal with non-scalar data
-        return Base.unsafe_load(ptr)
-    end
+    return copy ? Base.copy(result) : result
 end
 
-function unsafe_wrap(ptr, shape)
-    if length(shape) > 1
-        # We got a list of ptrs,
-        # but the first pointer points to the whole data
-        ptr = Base.unsafe_load(ptr)
+function _lammps_wrap(ptr::Ptr{<:Ptr{T}}, shape::NTuple{2}, copy=true) where T
+    ptr == C_NULL && error("Wrapping NULL-pointer!")
 
-        @assert length(shape) == 2
+    (count, ndata) = shape
 
-        # Note: Julia like Fortran is column-major
-        #       so the data is transposed from Julia's perspective
-        shape = reverse(shape)
-    end
+    ndata == 0 && return Matrix{T}(undef, count, ndata) # There is no data that can be wrapped
+
+    pointers = Base.unsafe_wrap(Array, ptr, ndata)
+
+    # This assert verifies that all the pointers are evenly spaced according to count.
+    # While it seems like this is allways the case, it's not explicitly stated in the
+    # API documentation. It also helps to verify, that the count is correct.
+    @assert all(diff(pointers) .== count*sizeof(T))
```

Comment:
This assert is quite costly. I would only do so in some sort of special debug mode.

```suggestion
    # @assert all(diff(pointers) .== count*sizeof(T))
```